### PR TITLE
Ability to set arbitrary custom data on Expenses

### PR DESCRIPTION
**Pull request by Betree**
Part of https://github.com/opencollective/opencollective/issues/6451

I've limited the max custom data size to 10kb for this first release.

### CODE REVIEW

**Comment by aerugo** [comment]
@Betree thanks! So, do understand correctly that this make it possible to call the expense submission flow with a customData parameter? And possible to query expenses with some customData?

**Comment by Betree** [comment]
> @Betree thanks! So, do understand correctly that this make it possible to call the expense submission flow with a customData parameter? And possible to query expenses with some customData?

Correct! We plan to plug this in the frontend via URL parameters, so that you can pass this custom data while reusing the existing form.

We also plan to add this field to the expenses search.

File: server/graphql/common/expenses.ts [hunk 1]
```diff
@@ -747,6 +758,23 @@ async function validateExpensePayout2FALimit(req, host, expense, expensePaidAmou
...
+
+  // Validate type: must be a JSON object
+  if (typeof value !== 'object') {
+    throw new ValidationFailed('Expense custom data must be an object');
+  }
+
+  // Validate size
+  const payloadSize = Buffer.byteLength(JSON.stringify(value), 'utf8');
+  if (payloadSize > 10e3) {
+    throw new ValidationFailed(`Expense custom data cannot exceed 10kb. Current size: ${payloadSize / 1000}kb`);
```
Comment (znarf) [inline server/graphql/common/expenses.ts:774]:
```suggestion
    throw new ValidationFailed(`Expense custom data cannot exceed 10kB. Current size: ${payloadSize / 1000}kB`);
```

A bit pedantic but here we go.

Comment (znarf) [reply server/graphql/common/expenses.ts:774]: https://en.wikipedia.org/wiki/Kilobyte

File: test/server/graphql/v2/mutation/ExpenseMutations.test.js [hunk 2]
```diff
@@ -401,6 +404,22 @@ describe('server/graphql/v2/mutation/ExpenseMutations', () => {
...
+      const expenseData = { ...getValidExpenseData(), payee: { legacyId: payee.id }, customData };
+      const collective = await fakeCollective();
+      const result = await graphqlQueryV2(
+        createExpenseMutation,
+        { expense: expenseData, account: { legacyId: collective.id } },
+        user,
+      );
+
+      expect(result.errors).to.exist;
+      expect(result.errors[0].message).to.eq('Expense custom data cannot exceed 10kb. Current size: 10.008kb');
```
Comment (znarf) [inline test/server/graphql/v2/mutation/ExpenseMutations.test.js:421]:
```suggestion
      expect(result.errors[0].message).to.eq('Expense custom data cannot exceed 10kB. Current size: 10.008kB');
```

**Review by znarf — APPROVED** [review]
Looks good.

**Review by gustavlrsn — APPROVED** [review]
Looks great!

**Pull request merged**